### PR TITLE
Add containerd graduation review proposal

### PR DESCRIPTION
**Pull request by estesp**
The submitted markdown document contains our proposal for graduation for the containerd CNCF project.

cc: @crosbymichael

Signed-off-by: Phil Estes <estesp@linux.vnet.ibm.com>

### CODE REVIEW

File: reviews/graduation-containerd.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+**_Docker engine_** - As noted in the opening paragraph, Docker continues to consume containerd as a key component within the Docker engine stack, and is actively working to remove Docker engine implementations where containerd implementations can be used as a replacement.
+
+With contributions from Microsoft and AWS, we believe that Azure and AWS are also looking at containerd for potential use within their public cloud offerings as well.
+
+**_Other Projects_** - While the above list provides a cross-section of well known uses of containerd, the simplicity and clear API layer for containerd has inspired many smaller projects around providing simple container management platforms. Two that have come from containerd community participants directly are Michael Crosby's [boss](https://github.com/crosbymichael/boss) project and Evan Hazlett's [stellar](https://github.com/ehazlett/stellar) project, as examples of higher layer functionality that can easily be built on the containerd base.
```
Comment (stevvooe) [inline reviews/graduation-containerd.md]: @ibuildthecloud also has a project.

Comment (estesp) [reply reviews/graduation-containerd.md]: Rio is listed above; unless he has another one. I wouldn't put it past him :)

File: reviews/graduation-containerd.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+**Have a public list of project adopters for at least the primary repository.**
+
+Containerd began life prior to its CNCF adoption as a lower-layer runtime manager for the Docker engine. Continuing today, containerd has widest usage and adoption as the layer between the Docker engine and the OCI runc executor. However, as containerd and its CRI plugin project have merged in January 2018, the combined use of containerd and the CRI plugin has grown to include several public cloud providers, as well as several projects who are attracted to the simplicity of the Go client API library for embedding container runtime capabilities.
+
+**_IBM Cloud Kubernetes Service (IKS)_** - offers containerd as the CRI runtime for v1.11 and, soon, v1.12.
+
+**_IBM Cloud Private (ICP)_** - IBM's on-premises cloud offering has containerd as a "tech preview" CRI runtime for the Kubernetes offered within this product for the past two releases, and plans to fully migrate to containerd in a future release.
+
+**_Google Cloud Kubernetes Engine (GKE)_** - offers containerd in "alpha clusters" on recent versions of Kubernetes.
```
Comment (AkihiroSuda) [inline reviews/graduation-containerd.md]: s/alpha/beta/ ? cc @random-liu

> cos_containerd OS images are available as a Beta feature in GKE v1.11 and higher.
 
https://cloud.google.com/kubernetes-engine/docs/concepts/using-containerd

File: reviews/graduation-containerd.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+At this time, we believe containerd is ready for the [graduation stage](https://github.com/cncf/toc/blob/master/process/graduation_criteria.adoc#graduation-stage) within the CNCF, and per the guidelines listed there, we detail our readiness in the section below.
+
+### CNCF Graduation Criteria
+
+**Committers from at least two organizations.**
+
+Containerd has had a variety of maintainers and reviewers since its inception, and currently have 14 committers representing NTT, Huawei, Docker, Google, IBM, Microsoft, Facebook, Tesla, and Cruise Automation. We also recognize **LGTM** rights for a group we call *reviewers*, of which there are currently five representing ZTE, Docker, and independants.
```
Comment (allencloud) [inline reviews/graduation-containerd.md]: I am wondering if containerd has a stable governing rules for maintainers. I think governing policy for maintainers' responsibility is quite healthy for the project, since I found that some members of containerd's maintainer list have not been active for a long time.

Comment (estesp) [reply reviews/graduation-containerd.md]: If I understand the comment correctly, you are right that we have some non-active maintainers. We were in the process of reaching out to them to understand their intention of continuing activity as a maintainer; the result of that is the PR that was just merged: https://github.com/containerd/project/pull/12

This does not change the make-up of the maintainer organization representation by much, but I will update the graduation PR to make it correct with the current status. Let me know if you had any other concerns here.

Comment (allencloud) [reply reviews/graduation-containerd.md]: Thanks for your update. Only this minor concern. Sincerely wish containerd could graduate soon.

File: reviews/graduation-containerd.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+**_Balena_** - Resin's [Balena](https://github.com/resin-os/balena) container engine, based on moby/moby but for edge, embedded, and IoT use cases, uses the containerd and runc stack in the same way that the Docker engine uses containerd.
+
+**_LinuxKit_** - the Moby project's [LinuxKit](https://github.com/linuxkit/linuxkit) for building secure, minimal Linux OS images in a container-native model uses containerd as the core runtime for system and service containers.
+
+**_BuildKit_** - The Moby project's [BuildKit](https://github.com/moby/buildkit) can use either runC or containerd as build execution backends for building container images. BuildKit support has also been built into the Docker engine in recent releases, making BuildKit provide the backend to the `docker build` command.
+
+**_Docker engine_** - As noted in the opening paragraph, Docker continues to consume containerd as a key component within the Docker engine stack, and is actively working to remove Docker engine implementations where containerd implementations can be used as a replacement.
```
Comment (resouer) [inline reviews/graduation-containerd.md]: Not sure if it's worth to mention but [katacontainers](https://github.com/kata-containers/runtime) is doing official integration with containerd shimv2 api. We are hoping to merge the [PR](https://github.com/kata-containers/runtime/pull/572) before KubeCon Seattle.

Comment (resouer) [reply reviews/graduation-containerd.md]: [Update] Fupan will try to get the shimv2 PR merged in 2 weeks :-)

Comment (resouer) [reply reviews/graduation-containerd.md]: cc @estesp WDYT?

**Comment by AkihiroSuda** [comment]
Can we also add https://github.com/firecracker-microvm/firecracker-containerd

File: reviews/graduation-containerd.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+**Adopt the CNCF Code of Conduct.**
+
+We have adopted the CNCF code of conduct and prominently make this clear in our [GOVERNANCE.md](https://github.com/containerd/project/blob/master/GOVERNANCE.md) document, which applies to all containerd projects and sub-projects.
+
+**Explicitly define a project governance and committer process.**
+
+Our project governance is clearly laid out in our [GOVERNANCE.md](https://github.com/containerd/project/blob/master/GOVERNANCE.md) document, including details on how to become a maintainer and how maintainer and contribution processes are handled. The maintainers for containerd, the CRI project, and all sub-projects are common, and maintained in our core project repo in the [MAINTAINERS](https://github.com/containerd/project/blob/master/MAINTAINERS) file.
+
+**Have a public list of project adopters for at least the primary repository.**
```
Comment (AkihiroSuda) [inline reviews/graduation-containerd.md]: Can we also add https://github.com/firecracker-microvm/firecracker-containerd to the adopter list

Comment (estesp) [reply reviews/graduation-containerd.md]: Added, thanks!

File: reviews/graduation-containerd.md (hunk 4 above)
Comment (estesp) [reply reviews/graduation-containerd.md]: Sounds great--just added Kata to the list of adopters!

**Comment by estesp** [comment]
Updated adopter list with Kata containers and AWS Firecracker VMM

**Comment by jovizhangwei** [comment]
@estesp can we also add Alibaba Cloud's ASK/ECI? thanks

The Alibaba Cloud Serverless Kubernetes [ASK&ECI](https://www.alibabacloud.com/en/products/elastic-container-instance) project uses containerd as its core lightweight runtime.

**Comment by estesp** [comment]
This is **not** a CNCF requirement for graduation, but may be of interest to reviewers to note that the containerd security audit provided via the CNCF is now published: https://github.com/containerd/containerd#security-audit

**Comment by caniszczyk** [comment]
final RFC from @cncf/toc and community before we kick a vote off next week

**Comment by bgrant0607** [comment]
Presentation: 
https://docs.google.com/presentation/d/1mf7_KX_EVTUf6QhGLNBKRvMRQMgOU7aK6CFhhYByg-0/edit#slide=id.g4481a3901f_1_0

Project health dashboard:
https://all.devstats.cncf.io/d/53/projects-health?orgId=1

File: reviews/graduation-containerd.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+At this time, we believe containerd is ready for the [graduation stage](https://github.com/cncf/toc/blob/master/process/graduation_criteria.adoc#graduation-stage) within the CNCF, and per the guidelines listed there, we detail our readiness in the section below.
+
+### CNCF Graduation Criteria
+
+**Committers from at least two organizations.**
+
+Containerd has had a variety of maintainers and reviewers since its inception, and currently have 12 committers representing Docker, NTT, Google, IBM, Microsoft, Facebook, Tesla, and Cruise Automation. We also recognize **LGTM** rights for a group we call *reviewers*, of which there are currently six reviewers representing ZTE, Huawei, Docker, Microsoft and an independent developer.
```
Comment (xiang90) [inline reviews/graduation-containerd.md]: There are 8 reviewers today: https://github.com/containerd/project/pull/13. The two additional reviewers represent Alibaba.

Comment (estesp) [reply reviews/graduation-containerd.md]: Thank you for the reminder! I will update the document @xiang90

**Comment by caniszczyk** [comment]
graduation vote is out https://lists.cncf.io/g/cncf-toc/message/2873

**Comment by caniszczyk** [comment]
Hey @estesp a question from @jbeda 

> I notice that the governance ultimate rolls up to the Moby TSC for ultimate arbitration of technical disputes: https://github.com/containerd/project/blob/master/GOVERNANCE.md. I'd love to see the project stand on its own vs. deferring to an outside community

**Comment by jbeda** [comment]
The graduation criteria simply states:
> Explicitly define a project governance and committer process. This preferably is laid out in a GOVERNANCE.md file and references an OWNERS.md file showing the current and emeritus committers.

This is obviously not the full criteria.  A GOVERNANCE.md that say "Joe decides everything" fulfills the criteria but obviously isn't acceptable. I read this more as create an **independent, fair and vendor neutral governance structure**.  The fact that the Moby TSC is the ultimate arbitrator means that we should be examining that governance structure too.  That is defined [here](https://github.com/moby/tsc).

Basically the Moby TSC is voted on from each of the participating projects.  Solomon Hykes also gets a vote. To really map the process here we'd have to dig into the individual projects.  There are no requirements or protections against vendor majority on the Moby TSC. By a quick look, 4 out of the 7 members of the Moby TSC are Docker employees. And the Moby TSC process could change in the future. 

I'd feel much much better about voting to graduate containerd if it had an independent governance structure.

**Comment by estesp** [comment]
Just to clarify some of the claims in the last comment: 
 - Moby TSC most definitely has protections against vendor-majority; specifically 1/3 of seats is max for any company (clearly stated in the "Committee Size" section). Only 2 of the current members are from Docker, but we have a timing issue in that Stefan just joined Docker recently and immediately notified us that he stepped down from the Moby TSC, but the site has not been updated. We have been internally discussing how to manage replacement of the seat, but that step has not occurred yet. We can fix the list with a PR however. Not sure who else was incorrectly numbered as a Docker employee.
 - Moby TSC is not beholden to Docker, if that is the implication, and the reason it exists was to separate governance of prior-Docker-owned projects to a neutral and fair governance process that is clearly defined on the TSC site. Many Kubernetes governance principles were the guide for the formation of the TSC. Several Kubernetes founders were involved in the creation and review of the TSC documents. Think of it as "step 1" in a long phase of discussions that led to removing the BDFL model from originally Docker-created projects.

I'm happy to respond to the broader discussion of whether Moby TSC is a proper "escape valve" for technical issues in containerd that cannot be resolved within the project itself. We have discussed it as a project recently, and feel fully free to use or not use the Moby TSC as an umbrella for governance, but haven't made any clear steps away from Moby TSC as a project to date. However I thought it would be useful to clarify some facts about the TSC before any rumor mills got wound up. :)

**Comment by jbeda** [comment]
I don't want to start any rumors so thanks for responding!

Sorry I missed the restriction on max representation of a single company.  I may be confused though as it looks like there are 4 ppl there from Docker (Justin, Stefan, Arnaud, Sebastian) and I would expect at max 2.  (I may have company affiliation wrong though as this stuff isn't always obvious).

All the same, I feel that we are on new ground here by having the technical arbitration for a project be outside of the project (and the CNCF).  I also realize that the need to use the Moby TSC will be rare to nonexistent and that this is currently more of a theoretical concern than a practical one. 

How hard would it be to cut that cord before we close the vote?

**Comment by crosbymichael** [comment]
From @estesp points, I think this addresses much of the issues you raised and I don't think this should impact the graduation vote at this point.  I'm not sure we want to cut the cord this close to the final vote unless others have issue with this.  I would like this to be a thought out and reviewed decision among the maintainers of the project and not done in haste unless this is going to impact all the graduation work we have put in so far.

As for going forward, I think this is something to bring up with the maintainers of the project and see of the Moby TSC is needed vs being totally independent.  This is something I will take on and start the discussion go with the maintainers of the project.  I think at this point in the project, the maturity of the project, we are able to move away from the Moby TSC but this is a decision for all the maintainers.

As far as who is at Docker on the TSC, with Stefan stepping down, we only have Justin and Sebastian, Arnaud is not at Docker.

**Comment by estesp** [comment]
> I may be confused though as it looks like there are currently 4 ppl there from Docker (Justin, Stefan, Arnaud, Sebastian) going to 3 (with Stefan stepping down) and I would expect at max 2. (I may have company affiliation wrong though as this stuff isn't always obvious).

- Arnaud left Docker almost 2 years ago; he is VP of Engineering for Vente-Privee, a European retail conglomerate (https://www.linkedin.com/in/aporterie/)
- It is our timing mistake to not have already updated `MEMBERS.md` to remove Stefan a few weeks ago when he emailed the current TSC members and formally stepped down from the TSC on the same day he joined Docker as an employee.
- That leaves Sebastiaan and Justin as the 2 representatives from Docker, Inc. on the current TSC.

**Comment by jbeda** [comment]
Ah -- I was confused as Arnaud's personal web page still has him working for Docker: https://icecrime.net/about/.  Thanks for the clarification.

**Comment by estesp** [comment]
@icecrime update your website! :joy:

**Comment by jbeda** [comment]
After talking to @estesp about how the project works and how the TSC connection *really is* a theoretical concern I agree that it is unreasonable to ask to change the governance at the 11th hour.  But I do encourage the project to look to remove this connection.  But that'll have to happen separate from this vote.

**Comment by crosbymichael** [comment]
@jbeda that sounds totally reasonable to me and we will get the discussion started with the containerd maintainers and community.

Thanks!

**Comment by bgrant0607** [comment]
@jbeda The question about the Moby TSC was helpful. I'm not sure it's a problem, but I found it odd when I noticed it recently.

However, we have no explicit policy regarding "protections against vendor majority". The most relevant criterion is "Have committers from at least two organizations."  I don't think we should impose a new "majority" policy on the fly for this specific graduation proposal. For example, would that imply that no project where 50% of the commits / contributions come from a single organization can graduate? That's not an issue for containerd, but would be an issue for at least 8 other projects. If we did want to impose such a policy, we should think about how we can help projects broaden their contributor bases, which perhaps we should do regardless.

**Comment by jbeda** [comment]
@bgrant0607 -- yes -- I'm not imposing new requirements here but just digging into the details of the whole governance structure of the project (which does import the Moby TSC).  We should follow up separately on updating the graduation requirements to something that is more meaningful than "has some governance".  Let's take the discussion off this issue.

**Comment by caniszczyk** [comment]
+1 binding TOC votes (8/9):
Jeff: https://lists.cncf.io/g/cncf-toc/message/2900
Joe: https://lists.cncf.io/g/cncf-toc/message/2914
Matt: https://lists.cncf.io/g/cncf-toc/message/2915
Xiang: https://lists.cncf.io/g/cncf-toc/message/2916
Brian: https://lists.cncf.io/g/cncf-toc/message/2924
Brendan: https://lists.cncf.io/g/cncf-toc/message/2931
Quinton: https://lists.cncf.io/g/cncf-toc/message/2933
Alexis: https://lists.cncf.io/g/cncf-toc/message/2934
